Including dividends

dimitrovmaksim opened the pull request:
Including csv file with dividends and showing total amount of earned dividends for each stock and average return per year in the main table

![capture](https://cloud.githubusercontent.com/assets/20456492/23103976/62379934-f6cd-11e6-9115-56b43092b8dd.PNG)

dimitrovmaksim committed:
Including csv file with dividends and showing total amount of earned
dividends for each stock and average return per year

antonrd commented on lib/compute_summaries.js:
```diff
@@ -3,20 +3,38 @@ var ComputeSummaries = {
+
+    for (var i = 0; i < dividends.length; i++){
+      var dividend = dividends[i];
+      this.dividends_details[dividend.payment_date] = dividend;
```
This array `dividends_details` is not needed I think, it's not used in the code besides here?

antonrd commented on lib/compute_summaries.js:
```diff
@@ -147,5 +194,9 @@ var ComputeSummaries = {
+  },
+
+  sortEvents: function(a,b) {
+    return new Date(a.date) - new Date(b.date);
```
I think these values (`a.date`, `b.date`) are already of type `Date` and there is no need to initialize new objects from them.

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
       }
+
+      if (trade.symbol in this.stock_dividends){
+        this.stock_dividends[trade.symbol].push(trade);
```
I would call this array something else, because it not only contains dividends, but also buys/sells. Maybe `stock_events` ?

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
+      this.stock_dividends[symbol].sort(this.sortEvents);
+      var stock_amount = 0;
+      var dividends_total = 0;
+      var event = this.stock_dividends[symbol];
```
This should probably be called `events` ?

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
+      var stock_amount = 0;
+      var dividends_total = 0;
+      var event = this.stock_dividends[symbol];
+      for (var i = 0; i < this.stock_dividends[symbol].length; i++){
```
Here you can instead have `i < events.length`

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
+          if (event[i].action == Constants.BUY) {
+            stock_amount += event[i].amount;
+          } else if (event[i].action == Constants.SELL){
+            stock_amount = stock_amount - event[i].amount;
```
This could be `stock_amount -= event[i].amount`

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
+      this.stock_dividends[symbol].average_return = average_return;
     }
 
+    // console.log(this.stock_summary);
```
Not needed

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
 
+    // console.log(this.stock_summary);
+
+    console.log(this.stock_dividends);
```
I think this print can be removed

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
+        }
+        days_passed = (this.today - this.stock_dividends[symbol][0].date) / (1000 * 60 * 60 * 24);
+        years = days_passed / 365;
+        average_return = dividends_total / years;
```
What if `years` is 0 ?

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +44,48 @@ var ComputeSummaries = {
+            dividends_total += stock_amount * event[i].dividend_amount;
+          }
+        }
+        days_passed = (this.today - this.stock_dividends[symbol][0].date) / (1000 * 60 * 60 * 24);
```
The first event could, in theory, be a dividend. Maybe we should take the first 'buy' and consider it as the beginning of the period.

antonrd commented on lib/compute_summaries.js:
```diff
@@ -3,20 +3,38 @@ var ComputeSummaries = {
+      this.dividends_details[dividend.payment_date] = dividend;
+      if ( dividend.symbol in this.stock_dividends) {
+        this.stock_dividends[dividend.symbol].push(this.dividends_details[dividend.payment_date]);
+      }
```
Can you add a message here to be printed if the symbol was not found in the array? This should indicate that a dividend for a company that is not in the portfolio was listed.

dimitrovmaksim committed:
Removed unneeded arrays. Renamed some of the arrays and variables with
more suitable name. Removed unnecessary console logs. Adjusted the code
to handle some edge cases.

antonrd commented on lib/compute_summaries.js:
```diff
@@ -3,20 +3,38 @@ var ComputeSummaries = {
     this.stock_details = {};
     this.total_portfolio_value = 0.0;
 
+    for (var i = 0; i < stocks.length; i++) {
```
The code how now grown a bit and it will be better if a few parts of this function are extracted to their own functions. This loop collecting the stock event could have its own function. Then the loop starting at line 29 could be in another function. And these would get called from here. They store a result in variables that are part of the object, so it will be possible to use this result further on.

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,11 +44,42 @@ var ComputeSummaries = {
-    for (var i = 0; i < stocks.length; i++) {
-      var stock = stocks[i];
-      this.stock_details[stock.symbol] = stock;
+    for (var symbol in this.stock_events){
```
this whole loop can also be in a separate function, which takes care of computing the dividend statistics

dimitrovmaksim committed:
Moved parts of the code into separated functions that are then called
inside the init function.

antonrd commented on lib/compute_summaries.js:
```diff
@@ -26,21 +31,21 @@ var ComputeSummaries = {
-    for (var i = 0; i < stocks.length; i++) {
-      var stock = stocks[i];
-      this.stock_details[stock.symbol] = stock;
+      if (trade.symbol in this.stock_events){
```
this could also go to `collectStockEvents`

antonrd commented on lib/compute_summaries.js:
```diff
@@ -147,5 +154,68 @@ var ComputeSummaries = {
+  computeAverageReturn: function(today, buy_events, dividends_total) {
+    days_passed = (today - buy_events[0].date) / (1000 * 60 * 60 * 24);
+    years = days_passed / 365;
+    if (years == 0 || years < 1 || dividends_total == 0) {
```
`years == 0` is covered by `years < 1`

antonrd commented on lib/compute_summaries.js:
```diff
@@ -147,5 +154,68 @@ var ComputeSummaries = {
+        this.computeAverageReturn(today, buy_events, dividends_total);
+      }
+      this.stock_events[symbol] = {dividends_total: dividends_total};
+      this.stock_events[symbol].average_return = average_return;
```
`average_return` now comes from nowhere

dimitrovmaksim committed:
Moved collecting the trade events to the collectStockEvents function and other small adjustements to the code.

antonrd commented on lib/compute_summaries.js:
```diff
@@ -147,5 +150,75 @@ var ComputeSummaries = {
+    days_passed = (today - buy_events[0].date) / (1000 * 60 * 60 * 24);
+    years = days_passed / 365;
+    if (years < 1 || dividends_total == 0) {
+      return average_return = 'N/A';
```
you can just `return 'N/A';`, same below

dimitrovmaksim committed:
Removed unnecessary variable declaration
